jupyterlab_wipp_plugin_creator/jupyterlab_wipp_plugin_creator/handlers.py

```diff
@@ -0,0 +1,128 @@
+
+# Create wipp plugin based on user input
+class CreatePlugin(WippHandler):
+    @tornado.web.authenticated
+    def post(self):
+        """
+        POST request handler,
+
+        Creates a temp folder with random name,
+        Generates plugin.json, requirements.txt, dockerfile,
+        Copies files
+        Builds dockerimage,
+
+        Input format:
+            {
+                'formdata':
+                    {
+                    'name': '',
+                    'version': '',
+                    'title': '',
+                    'description': ''
+                    }
+                'filepaths':
+                    string[]
+            }
+
+        """
+
+        pwd = os.getcwd()
+
+        # Random ID follows MongoDB format
+        randomId = gen_random_object_id()
+
+        pluginOutputPath = os.getenv("PLUGIN_TEMP_LOCATION")
+        # if ENV exists
+        if pluginOutputPath:
+            os.chdir(pluginOutputPath)
+            logger.info(f"ENV variable exists, output path set to {pluginOutputPath}.")
+            os.makedirs(f"{randomId}")
+            os.chdir(f"{randomId}")
+            pluginOutputPath = os.getcwd()
+            logger.info("Random folder name created: ", pluginOutputPath)
+
+        else:
+            logger.error("ENV variable doesn't exist, please use command 'export PLUGIN_TEMP_LOCATION = '...' to set. Terminating..")
+            quit()
+
+        # Read POST request
+        data = json.loads(self.request.body.decode("utf-8"))
+        form = data["formdata"]
+        filepaths = data["addedfilepaths"]
+        requirements = form["requirements"]
+
+        # Separate requirements key in the formdata form rest to write plugin.json and requirements.txt separately
+        form.pop("requirements")
+        form["containerId"] = "polusai/generated-plugins:" + randomId
+
+        # register plugin manifest to wipp CI
+        self.wipp.register_plugin(form)
+
+        # Generate files to temp folder
+        try:
+            with open("plugin.json", "w") as f1:
+                f1.write(json.dumps(form))
+            with open("requirements.txt", "w") as f2:
+                for req in requirements:
+                    f2.write(f"{req}\n")
+            # Read from Jinja2 template
+            os.chdir(pwd + '/jupyterlab_wipp_plugin_creator')
+            template = Template(open('dockerfile.j2').read())
+            # Generate dockerfile with user inputs, hardcoded for the time being
+            template.stream(userImage= "python").dump(pluginOutputPath + '/Dockerfile')
+            logger.info(f"Dockerfile Template generated from jinja2 template, src/dockerfile.j2" )
+
+
+        except Exception as e:
+            logger.error(f"Error writing files", exc_info=e)
+            # when error change back to root working dir
+            os.chdir(pwd)
+            self.write_error(500)
+
+        # Copy files to temp location with shutil
+        # Copy2 is like copy but preserves metadata
```

Comment:
Does it preserve the file ownership? Please, double check as it *might* cause some issues for docker build

Reply:
Great idea,
 in a red box at the top of [shutil's docs](https://docs.python.org/3/library/shutil.html) states:

**> Warning
> 
> Even the higher-level file copying functions (shutil.copy(), shutil.copy2()) cannot copy all file metadata.
> 
> On POSIX platforms, this means that file owner and group are lost as well as ACLs(access control lists I think). On Mac OS, the resource fork and other metadata are not used. This means that resources will be lost and file type and creator codes will not be correct. On Windows, file owners, ACLs and alternate data streams are not copied.
>** 

Seems a possible workaround: [Old stackoverflow](https://stackoverflow.com/questions/19787348/copy-file-keep-permissions-and-owner)

since then shutil has its own chown function which uses underlying os.chown

Reply:
Okay, probably is fine to keep it for now and revisit once Kubernetes/Argo/kaniko is added and *if* there are issues with permission from the kaniko container.